### PR TITLE
feat(responders): Raise a specialized subclass of HTTPNotFound

### PR DESCRIPTION
**Pull request by kgriffs**
When no route matches a request, raise a subclass of HTTPNotFound so that
a custom error handler can distinguish that specific case if desired.

(This was inspired by a recent chat with a user in the community.)

### CODE REVIEW

File: docs/api/errors.rst [hunk 1]
```diff
@@ -87,20 +87,107 @@ Mixins
 Predefined Errors
 -----------------
 
-.. automodule:: falcon
```
Comment (kgriffs) [inline docs/api/errors.rst]: This was preventing other docs from being able to create reference links to the error class docs.

**Comment by kgriffs** [comment]
Good ideas, I'll work on those.

**Comment by kgriffs** [comment]
> * We should adapt tests to verify the new behaviour, i.e. that the specialized subclass is raised, ideally in a scenario where the responder, if routed, would raise a plain `HTTPNotFound`.
> * `docs/api/routing.rst` needs to be revised
> * `docs/user/faq.rst` needs to be revised (`How can I write a custom handler for 404 and 500 pages in falcon?`)

@vytas7  I've attempted to address these points. Let me know what you think!

File: docs/user/faq.rst [hunk 2]
```diff
@@ -436,22 +436,23 @@ around:
...
-responder that simply raises an instance of :attr:`falcon.HTTPNotFound`. You
-can use :meth:`falcon.App.add_error_handler` to register a custom error handler
-for this exception type. Alternatively, you may be able to configure your web
-server to transform the response for you (e.g., using Nginx's ``error_page``
-directive).
+responder that simply raises an instance of :class:`~.HTTPRouteNotFound`, which
+the framework will in turn render as a 404 response. You can use
+:meth:`falcon.App.add_error_handler` to override the default handler for this
+exception type (or for its parent type,  :class:`~.HTTPNotFound`).
+Alternatively, you may be able to  configure your web server to transform the
```
Comment (vytas7) [inline docs/user/faq.rst]: Not that it matter much in reStructured Text, but there in an extraneous space after `able to`.

Comment (kgriffs) [reply docs/user/faq.rst]: fixed

**Pull request merged**